### PR TITLE
feat: IOx InfluxQL planner learns how to process time range expressions

### PR DESCRIPTION
**Pull request by stuartcarnie**
Helps #6112

Teach the InfluxQL planner how to identify and transform InfluxQL [time range][] expressions that appear in a `WHERE` clause, into a DataFusion logical expression. An example of a typical time range expression would be the following:

```sql
WHERE time > now() - 5s
```

The `time_range_to_df_expr` function is responsible for processing the expression such as `now() - 5s` and transforming it into a DataFusion logical expression. In the absence of an official specification, the rules applied by this function are determined from the Go InfluxQL parser and treated as the source of truth. Most of the implementation is sourced from the [`getTimeRange`][] and [`Reduce`][] functions. Any deviation from InfluxQL should be a superset, such that existing InfluxQL statements accepted by Cloud 2 and OSS should be processed successfully.

A time-range expression is determined when either the left or right hand side of a `ConditionalExpression` has a single node that refers to a `time` field. Whilst most of InfluxQL performs comparisons of fields using case-sensitive matches, this is a case-insensitive match, per the [`conditionExpr`][] function.

Binary expressions are treated as arithmetic expressions, with one exception when the left and right hand sides are strings, the expression is interpreted as a string concatenation operation. Ultimately, the resulting string must be a valid timestamp matching one of the following patterns.

A date string, where the time defaults to midnight:

```
YYYY-MM-DD
```

A date time string in the timezone specified by the `TZ` clause or UTC if unspecified:

```
YYYY-MM-DD hh:mm:ss.nnnnnnnnn
```

An RFC3339 + nanosecond date time string in the timezone specified by the string, ignoring the `TZ` clause.

Literal values interpreted as follows:

* single-quoted strings are interpreted as timestamps when either the left or right hand side of the binary expression is numeric.
* integer and float values as nanosecond offsets from the Unix epoch.
  * The Go implementation may interpret a number as a timestamp or duration, depending on context, however, in reality both are just offsets from the Unix epoch.

[time range]: https://docs.influxdata.com/influxdb/v1.8/query_language/explore-data/#absolute-time
[`getTimeRange`]: https://github.com/influxdata/influxql/blob/1ba470371ec093d57a726b143fe6ccbacf1b452b/ast.go#L5788-L5791
[`Reduce`]: https://github.com/influxdata/influxql/blob/1ba470371ec093d57a726b143fe6ccbacf1b452b/ast.go#L4850-L4852
[`conditionExpr`]: https://github.com/influxdata/influxql/blob/1ba470371ec093d57a726b143fe6ccbacf1b452b/ast.go#L5751-L5756
[`TZ`]: https://docs.influxdata.com/influxdb/v1.8/query_language/explore-data/#the-time-zone-clause

### CODE REVIEW

File: influxdb_influxql_parser/Cargo.toml [hunk 1]
```diff
@@ -10,9 +10,11 @@ nom = { version = "7", default-features = false, features = ["std"] }
 once_cell = "1"
 chrono = { version = "0.4", default-features = false, features = ["std"] }
 chrono-tz = { version = "0.8" }
-workspace-hack = { path = "../workspace-hack" }
+num-traits = "0.2"
+workspace-hack = { path = "../workspace-hack"}
+#workspace-hack = { path = "../workspace-hack" }
```
Comment (crepererum) [inline influxdb_influxql_parser/Cargo.toml:15]:
```suggestion
```

File: iox_query/src/plan/influxql/timestamp.rs [hunk 2]
```diff
@@ -0,0 +1,190 @@
...
+}
+
+#[cfg(test)]
+mod test {
+    use super::parse_timestamp;
+
+    #[test]
+    fn test_parse_timestamp() {
+        use chrono_tz::America::New_York; // a timezone in the Western hemisphere
+        use chrono_tz::Australia::Hobart; // a timezone in the Eastern hemisphere
```
Comment (alamb) [inline iox_query/src/plan/influxql/timestamp.rs:80]: I love the 🥚  with the author's location in tests 👍

File: influxdb_influxql_parser/src/expression/conditional.rs [hunk 3]
```diff
@@ -185,6 +197,16 @@ pub(crate) fn conditional_expression(i: &str) -> ParseResult<&str, ConditionalEx
     disjunction(i)
 }
 
+impl FromStr for ConditionalExpression {
+    type Err = ();
```
Comment (alamb) [inline influxdb_influxql_parser/src/expression/conditional.rs:201]: I wonder why not return the underlying error 

```suggestion
    type Err = &str;
```

And then I think you could avoid the `map_err` below 🤔

File: influxdb_influxql_parser/src/literal.rs [hunk 4]
```diff
@@ -99,7 +103,7 @@ impl From<Regex> for Literal {
 impl Display for Literal {
     fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
         match self {
-            Self::Integer(v) => write!(f, "{}", v),
+            Self::Integer(v) => Display::fmt(v, f),
```
Comment (alamb) [inline influxdb_influxql_parser/src/literal.rs:106]: is this change needed? I don't think it is wrong, but it is surprising and inconsistent with the rest of this function which uses `write!`

File: iox_query/src/plan/influxql.rs [hunk 5]
```diff
@@ -1,695 +1,12 @@
...
 mod field_mapper;
+mod planner;
+mod planner_time_range_expression;
 mod rewriter;
 mod test_utils;
+mod timestamp;
+mod util;
 mod var_ref;
 
-use crate::plan::influxql::rewriter::rewrite_statement;
```
Comment (alamb) [inline iox_query/src/plan/influxql.rs:8]: this code is mostly moved into its own module, right?

File: iox_query/src/plan/influxql/planner.rs [hunk 6]
```diff
@@ -0,0 +1,769 @@
...
+        //
+
+        //
+        // Scenarios: Multiple measurements, including
+        //
+        // * explicitly specified,
+        // * regular expression matching
+    }
+
+    /// This module contains esoteric features of InfluxQL that are identified during
```
Comment (alamb) [inline iox_query/src/plan/influxql/planner.rs:718]: 👍

File: iox_query/src/plan/influxql/planner_time_range_expression.rs [hunk 7]
```diff
@@ -0,0 +1,542 @@
...
+        DFExpr::Literal(ScalarValue::Int64(Some(rhs))) => {
+            op.try_reduce(lhs, *rhs).ok_or_else(|| {
+                DataFusionError::Plan("invalid operator for float expression".to_string())
+            })?
+        }
+        _ => return Err(DataFusionError::Plan("invalid float expression".into())),
+    }))
+}
+
+///
```
Comment (alamb) [inline iox_query/src/plan/influxql/planner_time_range_expression.rs:277]: seems like this might have meant to be a comment

File: iox_query/src/plan/influxql/planner_time_range_expression.rs [hunk 8]
```diff
@@ -0,0 +1,542 @@
...
+            "TimestampNanosecond(1157082309000000000, None)"
+        );
+
+        // nested evaluation order
+        assert_expr!(
+            "now() - (6m - (1m * 5))",
+            r#"now() - IntervalMonthDayNano("60000000000")"#
+        );
+
+        // Fallible
```
Comment (alamb) [inline iox_query/src/plan/influxql/planner_time_range_expression.rs:496]: 👍  for error testing

File: iox_query/src/plan/influxql/timestamp.rs [hunk 9]
```diff
@@ -0,0 +1,190 @@
+use chrono::{DateTime, FixedOffset, NaiveDate, NaiveDateTime, NaiveTime, Offset, TimeZone};
+use datafusion::common::{DataFusionError, Result};
+
+/// Parse the timestamp string and return a DateTime in UTC.
+fn parse_timestamp_utc(s: &str) -> Result<DateTime<FixedOffset>> {
+    // 1a. Try a date time format string with nanosecond precision and then without
+    //    https://github.com/influxdata/influxql/blob/1ba470371ec093d57a726b143fe6ccbacf1b452b/ast.go#L3661
+    NaiveDateTime::parse_from_str(s, "%Y-%m-%d %H:%M:%S%.f")
+        // 1b. Try a date time format string without nanosecond precision
```
Comment (alamb) [inline iox_query/src/plan/influxql/timestamp.rs:9]: This is very similar to the arrow-rs implementation https://docs.rs/arrow-cast/32.0.0/src/arrow_cast/parse.rs.html#72 (but support different formats, as it should) 👍

File: influxdb_influxql_parser/src/expression/arithmetic.rs [hunk 10]
```diff
@@ -276,6 +277,64 @@ pub enum BinaryOperator {
     BitwiseXor,
 }
 
+trait Reduce<Rhs = Self> {
```
Comment (alamb) [inline influxdb_influxql_parser/src/expression/arithmetic.rs:280]: I think `reduce` is just fine to do a conversion though it seems like it may be  awkard sometimes because Output and Self aren't the same type, so you have to convert back from Exprs sometimes

Another approach would be to do a straight forward recursive pattern match down the tree like

```
trait ToExpr {
  fn to_expr(&self) -> Result<Expr>
}

impl ToExpr for BinaryOperator {
  fn to_expr(&self) -> {
    binary_op(lhs.to_expr(), get_df_operator(self.op), rhs.to_expr())
  }
}
````

**Review by alamb — APPROVED** [review]
I can't really review if this faithfully implements all of the InfluxQL semantics, but I did review all the code, and found it easy to read and understand, as well as nicely commented and tested. 🏆

File: influxdb_influxql_parser/Cargo.toml (hunk 1 above)
Comment (stuartcarnie) [reply influxdb_influxql_parser/Cargo.toml:15]: Ugh, sorry about that 🙄. I hadn't realised that change was pulled in to this PR. I use it locally to build the parser in my InfluxQL analyser binary, as it builds much faster when it doesn't have to pull in a lot of unnecessary packages.

File: influxdb_influxql_parser/src/literal.rs (hunk 4 above)
Comment (stuartcarnie) [reply influxdb_influxql_parser/src/literal.rs:106]: Not sure why I didn't change them all – will revert that change for now.

File: influxdb_influxql_parser/src/expression/arithmetic.rs (hunk 10 above)
Comment (stuartcarnie) [reply influxdb_influxql_parser/src/expression/arithmetic.rs:280]: I should have deleted the `Reduce` trait, as it is no longer used. `clippy` doesn't give you any indication that a trait is unused 🤔

File: influxdb_influxql_parser/src/expression/conditional.rs (hunk 3 above)
Comment (stuartcarnie) [reply influxdb_influxql_parser/src/expression/conditional.rs:201]: I'll do better still and return a `ParseError`, consistent with the `parse_statements` public function. Thanks for pointing that one out!

Comment (stuartcarnie) [reply influxdb_influxql_parser/src/expression/conditional.rs:201]: Better yet, I've introduced a `parse_conditional_expression` public API that, similar to the `parse_statements`, consumes the entire input to ensure the input is a valid `ConditionalExpression`. Like the `parse_statements` function, `parse_conditional_expression` returns a `ParseError` for any failures. The `parse_conditional_expression` function is used to implement the `FromStr` trait.

File: influxdb_influxql_parser/src/lib.rs [hunk 11]
```diff
@@ -19,7 +19,7 @@ use crate::internal::Error as InternalError;
 use crate::statement::{statement, Statement};
 use nom::combinator::eof;
 use nom::Offset;
-use std::fmt::{Debug, Display, Formatter};
+use common::ParseError;
```
Comment (stuartcarnie) [inline influxdb_influxql_parser/src/lib.rs:22]: Moved to the `common` module, so it is more obvious the `ParseError` type is shared across modules.

File: influxdb_influxql_parser/src/literal.rs (hunk 4 above)
Comment (stuartcarnie) [reply influxdb_influxql_parser/src/literal.rs]: Done

File: iox_query/src/plan/influxql.rs (hunk 5 above)
Comment (stuartcarnie) [reply iox_query/src/plan/influxql.rs:8]: Yes, correct, it's all in the `influxql` module now. If it is worth while, we could move it all to another crate in a future PR.

File: iox_query/src/plan/influxql/planner_time_range_expression.rs (hunk 7 above)
Comment (stuartcarnie) [reply iox_query/src/plan/influxql/planner_time_range_expression.rs:289]: Thanks – I've added additional comments, where useful.
